File: repeat_py/63.py
from typing import List


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        rows = len(obstacleGrid)
        cols = len(obstacleGrid[0])
        mem = [[0 for c in range(cols)] for r in range(rows)]
        # mem = [i[:] for i in [[0] * cols] * rows]
        mem[0][0] = 1 if obstacleGrid[0][0] == 0 else 0
        for r in range(rows):
            for c in range(cols):
                if r == 0 and c == 0:
                    continue
                elif obstacleGrid[r][c] == 1:
                    mem[r][c] = 0
                else:
                    up = 0 if r == 0 else mem[r - 1][c]
                    left = 0 if c == 0 else mem[r][c - 1]
                    mem[r][c] = up + left
        return mem[rows - 1][cols - 1]


class SolutionMemoryCompression:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        rows = len(obstacleGrid)
        cols = len(obstacleGrid[0])
        mem = [0] * cols
        for r in range(rows):
            for c in range(cols):
                if obstacleGrid[r][c] == 1:
                    mem[c] = 0
                else:
                    up = 0 if r == 0 else mem[c]
                    left = 0 if c == 0 else mem[c - 1]
                    mem[c] = up + left if (c != 0 or r != 0) else 1
        return mem[cols - 1]

File: dp/63.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) { // use grid as a dp-table
        int rc = obstacleGrid.length;
        int cc = obstacleGrid[0].length;
        for (int r = 0; r < rc; r++){
            for (int c = 0; c < cc; c++){
                if (obstacleGrid[r][c] == 1){
                    obstacleGrid[r][c] = 0;
                } else {
                    int upPathCount = r == 0 ? 0 : obstacleGrid[r - 1][c];
                    int leftPathCount = c == 0 ? 0 : obstacleGrid[r][c - 1];
                    obstacleGrid[r][c] = r == 0 && c == 0 ? 1 : upPathCount + leftPathCount;
                }
            }
        }
        return obstacleGrid[rc - 1][cc - 1];
    }
}

class Solution2 { // without modifying grid
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int rc = obstacleGrid.length;
        int cc = obstacleGrid[0].length;
        int[] mem = new int[cc];
        for (int r = 0; r < rc; r++){
            for (int c = 0; c < cc; c++){
                if (obstacleGrid[r][c] == 1){
                    mem[c] = 0;
                } else {
                    int upPathCount = r == 0 ? 0 : mem[c];
                    int leftPathCount = c == 0 ? 0 : mem[c - 1];
                    mem[c] = r == 0 && c == 0 ? 1 : upPathCount + leftPathCount;
                }
            }
        }
        return mem[cc - 1];
    }
}
